### PR TITLE
Next App Router: Support Meta tags using new generateMetadata api

### PR DESCRIPTION
**Pull request by nsams**
This adds support for meta data with our Seo block using the generateMedadata next api.

- add support in graphQLFetch to use `method: "GET"` to enable request memoization (else the request would be made twice per site request)

TODO (later?):

- [ ] move helper function that creates `Metadata` out of block data into library
- [ ] <s>support passtedTitle (needed? simply overwrite return of helper)</s>
- [ ] <s>support passedCanonicalUrl (needed? simply overwrite return of helper - it's also in of openGraph.url - why?)</s>
- [x] support openGraph.image
- [x] support structuredData (json-ld) (must be part of Page component!)

### CODE REVIEW

File: demo/site/src/app/[[...path]]/page.tsx [hunk 1]
```diff
@@ -28,10 +33,49 @@ export default async function Page({ params }: { params: { path: string[] } }) {
...
-    const data = await graphqlFetch<GQLDocumentTypeQuery, GQLDocumentTypeQueryVariables>(documentTypeQuery, {
-        path: `/${(params.path ?? []).join("/")}`,
+    const data = await graphqlFetch<GQLDocumentTypeQuery, GQLDocumentTypeQueryVariables>(
+        documentTypeQuery,
+        {
+            path: `/${(params.path ?? []).join("/")}`,
+            scope,
+        },
+        { method: "GET" }, //for request memoization
+    );
```
Comment (johnnyomair) [inline demo/site/src/app/[[...path]]/page.tsx:43]: Should we move this into a `fetchPageTreeNode` function and reuse it?

File: demo/site/src/documentTypes/Page.tsx [hunk 2]
```diff
@@ -42,21 +43,75 @@ const pageQuery = gql`
...
+    // TODO move into library
+    return {
+        title: document.seo.htmlTitle || data.pageContent.name, //support passedTitle
+        description: document.seo.metaDescription,
+        openGraph: {
+            title: document.seo.openGraphTitle,
+            description: document.seo.openGraphDescription,
+            type: "website",
+            url: document.seo.canonicalUrl, //support passedCanonicalUrl
+            // TODO image: generateImageUrl({ src: openGraphImageUrlTemplate, width: 1024 }, 1 / 1),
```
Comment (johnnyomair) [inline demo/site/src/documentTypes/Page.tsx:96]: Why don't you add this right in this PR?

**Comment by johnnyomair** [comment]
I don't think we need overriding the fields using passed* options.

File: demo/site/src/app/[[...path]]/page.tsx (hunk 1 above)
Comment (nsams) [reply demo/site/src/app/[[...path]]/page.tsx]: done in 9e7d81c4bf67983f5d493406f40974a3f22385ae

I'm not sure if it's really a good idea as it requires 3 arguments

Comment (nsams) [reply demo/site/src/app/[[...path]]/page.tsx]: refactored again in 7f6d026b23cd7d922602e1bfb6435a8e7f561c7c

still not ideal, scope is also needed outside

Comment (johnnyomair) [reply demo/site/src/app/[[...path]]/page.tsx]: > I'm not sure if it's really a good idea as it requires 3 arguments

You can also revert. I'd rather have duplicate code than a hasty abstraction. Haven't thought it through when making the suggestion.

**Review by johnnyomair — APPROVED** [review]

**Comment by johnnyomair** [comment]
@raphaelblum please review.

**Comment by johnnyomair** [comment]
> move helper function that creates Metadata out of block data into library

Do you want to do this in this PR or in a follow-up PR? Should we wait until we've gained some experience in projects?

**Comment by nsams** [comment]
> Should we wait until we've gained some experience in projects?

yes

File: demo/site/src/documentTypes/Page.tsx [hunk 3]
```diff
@@ -41,18 +42,77 @@ const pageQuery = gql`
...
+            url: document.seo.canonicalUrl,
+            images: document.seo.openGraphImage.block?.urlTemplate
+                ? generateImageUrl({ src: document.seo.openGraphImage.block?.urlTemplate, width: 1024 }, 1 / 1)
+                : undefined,
+        },
+        robots: {
+            index: !document.seo.noIndex,
+        },
+        alternates: {
+            canonical: document.seo.canonicalUrl,
```
Comment (raphaelblum) [inline demo/site/src/documentTypes/Page.tsx:100]: As far as i know, if there is no canonicalUrl set in the CMS we should reference to the own page.

Every time we didn't have it, SEO agencies told us that it is important.

Screenshot 2024-05-02 at 10 22 08

Comment (nsams) [reply demo/site/src/documentTypes/Page.tsx]: I added it, although we don't have the site url env anymore - that will have to wait for the new site config

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**